File: front/lib/domain/sync/entity_payload.dart
import 'package:amap_en_ligne/domain/model/admin_member_join_request.dart';
import 'package:amap_en_ligne/domain/model/admin_organization_request.dart';
import 'package:amap_en_ligne/domain/model/admin_producer_request.dart';
import 'package:amap_en_ligne/domain/model/attendance_email_request.dart';
import 'package:amap_en_ligne/domain/model/basket_exchange.dart';
import 'package:amap_en_ligne/domain/model/contract.dart';
import 'package:amap_en_ligne/domain/model/delivery_template.dart';
import 'package:amap_en_ligne/domain/model/device_token.dart';
import 'package:amap_en_ligne/domain/model/error_report.dart';
import 'package:amap_en_ligne/domain/model/member.dart';
import 'package:amap_en_ligne/domain/model/member_invitation.dart';
import 'package:amap_en_ligne/domain/model/notification.dart';
import 'package:amap_en_ligne/domain/model/organization.dart';
import 'package:amap_en_ligne/domain/model/owner.dart';
import 'package:amap_en_ligne/domain/model/owner_invitation.dart';
import 'package:amap_en_ligne/domain/model/producer_account.dart';
import 'package:amap_en_ligne/domain/model/product_type.dart';
import 'package:amap_en_ligne/domain/sync/entity_type.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'entity_payload.freezed.dart';

/// Polymorphic carrier for the typed body of an entity in the sync protocol.
///
/// Wire `type` discriminator matches the corresponding `EntityType` constant
/// on the back. Hand-rolled (rather than Freezed unions) because Freezed 3.x
/// does not emit the discriminator on `toJson` for single-variant unions, and
/// we still want a sealed type for exhaustive `switch`.
sealed class EntityPayload {
  const EntityPayload();

  factory EntityPayload.fromJson(Map<String, dynamic> json) =>
      switch (json['type']) {
        'ProductType' => ProductTypePayload.fromJson(json),
        'Organization' => OrganizationPayload.fromJson(json),
        'ProducerAccount' => ProducerAccountPayload.fromJson(json),
        'Member' => MemberPayload.fromJson(json),
        'MemberJoinRequest' => MemberJoinRequestPayload.fromJson(json),
        'Contract' => ContractPayload.fromJson(json),
        'DeliveryTemplate' => DeliveryTemplatePayload.fromJson(json),
        'OrganizationRequest' => OrganizationRequestPayload.fromJson(json),
        'ProducerRequest' => ProducerRequestPayload.fromJson(json),
        'Owner' => OwnerPayload.fromJson(json),
        'MemberInvitation' => MemberInvitationPayload.fromJson(json),
        'OwnerInvitation' => OwnerInvitationPayload.fromJson(json),
        'BasketExchange' => BasketExchangePayload.fromJson(json),
        'Notification' => NotificationPayload.fromJson(json),
        'DeviceToken' => DeviceTokenPayload.fromJson(json),
        'AttendanceEmailRequest' => AttendanceEmailRequestPayload.fromJson(
          json,
        ),
        'ErrorReport' => ErrorReportPayload.fromJson(json),
        final t => throw FormatException('Unknown EntityPayload type: $t'),
      };

  EntityType get entityType;

  Map<String, dynamic> toJson();
}

@Freezed(toJson: false, fromJson: false)
abstract class ProductTypePayload extends EntityPayload
    with _$ProductTypePayload {
  const factory ProductTypePayload({required ProductType productType}) =
      _ProductTypePayload;
  const ProductTypePayload._() : super();

  factory ProductTypePayload.fromJson(Map<String, dynamic> json) =>
      ProductTypePayload(
        productType: ProductType.fromJson(
          json['productType'] as Map<String, dynamic>,
        ),
      );

  @override
  EntityType get entityType => EntityType.productType;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'ProductType',
    'productType': productType.toJson(),
  };
}

@Freezed(toJson: false, fromJson: false)
abstract class OrganizationPayload extends EntityPayload
    with _$OrganizationPayload {
  const factory OrganizationPayload({required Organization organization}) =
      _OrganizationPayload;
  const OrganizationPayload._() : super();

  factory OrganizationPayload.fromJson(Map<String, dynamic> json) =>
      OrganizationPayload(
        organization: Organization.fromJson(
          json['organization'] as Map<String, dynamic>,
        ),
      );

  @override
  EntityType get entityType => EntityType.organization;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'Organization',
    'organization': organization.toJson(),
  };
}

@Freezed(toJson: false, fromJson: false)
abstract class ProducerAccountPayload extends EntityPayload
    with _$ProducerAccountPayload {
  const factory ProducerAccountPayload({
    required ProducerAccount producerAccount,
  }) = _ProducerAccountPayload;
  const ProducerAccountPayload._() : super();

  factory ProducerAccountPayload.fromJson(Map<String, dynamic> json) =>
      ProducerAccountPayload(
        producerAccount: ProducerAccount.fromJson(
          json['producerAccount'] as Map<String, dynamic>,
        ),
      );

  @override
  EntityType get entityType => EntityType.producerAccount;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'ProducerAccount',
    'producerAccount': producerAccount.toJson(),
  };
}

@Freezed(toJson: false, fromJson: false)
abstract class MemberPayload extends EntityPayload with _$MemberPayload {
  const factory MemberPayload({required Member member}) = _MemberPayload;
  const MemberPayload._() : super();

  factory MemberPayload.fromJson(Map<String, dynamic> json) => MemberPayload(
    member: Member.fromJson(json['member'] as Map<String, dynamic>),
  );

  @override
  EntityType get entityType => EntityType.member;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'Member',
    'member': member.toJson(),
  };
}

@Freezed(toJson: false, fromJson: false)
abstract class MemberJoinRequestPayload extends EntityPayload
    with _$MemberJoinRequestPayload {
  const factory MemberJoinRequestPayload({
    required AdminMemberJoinRequest memberJoinRequest,
  }) = _MemberJoinRequestPayload;
  const MemberJoinRequestPayload._() : super();

  factory MemberJoinRequestPayload.fromJson(Map<String, dynamic> json) =>
      MemberJoinRequestPayload(
        memberJoinRequest: AdminMemberJoinRequest.fromJson(
          json['memberJoinRequest'] as Map<String, dynamic>,
        ),
      );

  @override
  EntityType get entityType => EntityType.memberJoinRequest;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'MemberJoinRequest',
    'memberJoinRequest': memberJoinRequest.toJson(),
  };
}

@Freezed(toJson: false, fromJson: false)
abstract class ContractPayload extends EntityPayload with _$ContractPayload {
  const factory ContractPayload({required Contract contract}) =
      _ContractPayload;
  const ContractPayload._() : super();

  factory ContractPayload.fromJson(Map<String, dynamic> json) =>
      ContractPayload(
        contract: Contract.fromJson(json['contract'] as Map<String, dynamic>),
      );

  @override
  EntityType get entityType => EntityType.contract;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'Contract',
    'contract': contract.toJson(),
  };
}

@Freezed(toJson: false, fromJson: false)
abstract class DeliveryTemplatePayload extends EntityPayload
    with _$DeliveryTemplatePayload {
  const factory DeliveryTemplatePayload({
    required DeliveryTemplate deliveryTemplate,
  }) = _DeliveryTemplatePayload;
  const DeliveryTemplatePayload._() : super();

  factory DeliveryTemplatePayload.fromJson(Map<String, dynamic> json) =>
      DeliveryTemplatePayload(
        deliveryTemplate: DeliveryTemplate.fromJson(
          json['deliveryTemplate'] as Map<String, dynamic>,
        ),
      );

  @override
  EntityType get entityType => EntityType.deliveryTemplate;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'DeliveryTemplate',
    'deliveryTemplate': deliveryTemplate.toJson(),
  };
}

@Freezed(toJson: false, fromJson: false)
abstract class OrganizationRequestPayload extends EntityPayload
    with _$OrganizationRequestPayload {
  const factory OrganizationRequestPayload({
    required AdminOrganizationRequest organizationRequest,
  }) = _OrganizationRequestPayload;
  const OrganizationRequestPayload._() : super();

  factory OrganizationRequestPayload.fromJson(Map<String, dynamic> json) =>
      OrganizationRequestPayload(
        organizationRequest: AdminOrganizationRequest.fromJson(
          json['organizationRequest'] as Map<String, dynamic>,
        ),
      );

  @override
  EntityType get entityType => EntityType.organizationRequest;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'OrganizationRequest',
    'organizationRequest': organizationRequest.toJson(),
  };
}

@Freezed(toJson: false, fromJson: false)
abstract class ProducerRequestPayload extends EntityPayload
    with _$ProducerRequestPayload {
  const factory ProducerRequestPayload({
    required AdminProducerRequest producerRequest,
  }) = _ProducerRequestPayload;
  const ProducerRequestPayload._() : super();

  factory ProducerRequestPayload.fromJson(Map<String, dynamic> json) =>
      ProducerRequestPayload(
        producerRequest: AdminProducerRequest.fromJson(
          json['producerRequest'] as Map<String, dynamic>,
        ),
      );

  @override
  EntityType get entityType => EntityType.producerRequest;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'ProducerRequest',
    'producerRequest': producerRequest.toJson(),
  };
}

@Freezed(toJson: false, fromJson: false)
abstract class OwnerPayload extends EntityPayload with _$OwnerPayload {
  const factory OwnerPayload({required Owner owner}) = _OwnerPayload;
  const OwnerPayload._() : super();

  factory OwnerPayload.fromJson(Map<String, dynamic> json) => OwnerPayload(
    owner: Owner.fromJson(json['owner'] as Map<String, dynamic>),
  );

  @override
  EntityType get entityType => EntityType.owner;

  @override
  Map<String, dynamic> toJson() => {'type': 'Owner', 'owner': owner.toJson()};
}

@Freezed(toJson: false, fromJson: false)
abstract class MemberInvitationPayload extends EntityPayload
    with _$MemberInvitationPayload {
  const factory MemberInvitationPayload({
    required MemberInvitation memberInvitation,
  }) = _MemberInvitationPayload;
  const MemberInvitationPayload._() : super();

  factory MemberInvitationPayload.fromJson(Map<String, dynamic> json) =>
      MemberInvitationPayload(
        memberInvitation: MemberInvitation.fromJson(
          json['memberInvitation'] as Map<String, dynamic>,
        ),
      );

  @override
  EntityType get entityType => EntityType.memberInvitation;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'MemberInvitation',
    'memberInvitation': memberInvitation.toJson(),
  };
}

@Freezed(toJson: false, fromJson: false)
abstract class OwnerInvitationPayload extends EntityPayload
    with _$OwnerInvitationPayload {
  const factory OwnerInvitationPayload({
    required OwnerInvitation ownerInvitation,
  }) = _OwnerInvitationPayload;
  const OwnerInvitationPayload._() : super();

  factory OwnerInvitationPayload.fromJson(Map<String, dynamic> json) =>
      OwnerInvitationPayload(
        ownerInvitation: OwnerInvitation.fromJson(
          json['ownerInvitation'] as Map<String, dynamic>,
        ),
      );

  @override
  EntityType get entityType => EntityType.ownerInvitation;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'OwnerInvitation',
    'ownerInvitation': ownerInvitation.toJson(),
  };
}

@Freezed(toJson: false, fromJson: false)
abstract class BasketExchangePayload extends EntityPayload
    with _$BasketExchangePayload {
  const factory BasketExchangePayload({
    required BasketExchange basketExchange,
  }) = _BasketExchangePayload;
  const BasketExchangePayload._() : super();

  factory BasketExchangePayload.fromJson(Map<String, dynamic> json) =>
      BasketExchangePayload(
        basketExchange: BasketExchange.fromJson(
          json['basketExchange'] as Map<String, dynamic>,
        ),
      );

  @override
  EntityType get entityType => EntityType.basketExchange;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'BasketExchange',
    'basketExchange': basketExchange.toJson(),
  };
}

@Freezed(toJson: false, fromJson: false)
abstract class NotificationPayload extends EntityPayload
    with _$NotificationPayload {
  const factory NotificationPayload({required AppNotification notification}) =
      _NotificationPayload;
  const NotificationPayload._() : super();

  factory NotificationPayload.fromJson(Map<String, dynamic> json) =>
      NotificationPayload(
        notification: AppNotification.fromJson(
          json['notification'] as Map<String, dynamic>,
        ),
      );

  @override
  EntityType get entityType => EntityType.notification;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'Notification',
    'notification': notification.toJson(),
  };
}

@Freezed(toJson: false, fromJson: false)
abstract class DeviceTokenPayload extends EntityPayload
    with _$DeviceTokenPayload {
  const factory DeviceTokenPayload({required DeviceToken deviceToken}) =
      _DeviceTokenPayload;
  const DeviceTokenPayload._() : super();

  factory DeviceTokenPayload.fromJson(Map<String, dynamic> json) =>
      DeviceTokenPayload(
        deviceToken: DeviceToken.fromJson(
          json['deviceToken'] as Map<String, dynamic>,
        ),
      );

  @override
  EntityType get entityType => EntityType.deviceToken;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'DeviceToken',
    'deviceToken': deviceToken.toJson(),
  };
}

@Freezed(toJson: false, fromJson: false)
abstract class AttendanceEmailRequestPayload extends EntityPayload
    with _$AttendanceEmailRequestPayload {
  const factory AttendanceEmailRequestPayload({
    required AttendanceEmailRequest attendanceEmailRequest,
  }) = _AttendanceEmailRequestPayload;
  const AttendanceEmailRequestPayload._() : super();

  factory AttendanceEmailRequestPayload.fromJson(Map<String, dynamic> json) =>
      AttendanceEmailRequestPayload(
        attendanceEmailRequest: AttendanceEmailRequest.fromJson(
          json['attendanceEmailRequest'] as Map<String, dynamic>,
        ),
      );

  @override
  EntityType get entityType => EntityType.attendanceEmailRequest;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'AttendanceEmailRequest',
    'attendanceEmailRequest': attendanceEmailRequest.toJson(),
  };
}

@Freezed(toJson: false, fromJson: false)
abstract class ErrorReportPayload extends EntityPayload
    with _$ErrorReportPayload {
  const factory ErrorReportPayload({required ErrorReport errorReport}) =
      _ErrorReportPayload;
  const ErrorReportPayload._() : super();

  factory ErrorReportPayload.fromJson(Map<String, dynamic> json) =>
      ErrorReportPayload(
        errorReport: ErrorReport.fromJson(
          json['errorReport'] as Map<String, dynamic>,
        ),
      );

  @override
  EntityType get entityType => EntityType.errorReport;

  @override
  Map<String, dynamic> toJson() => {
    'type': 'ErrorReport',
    'errorReport': errorReport.toJson(),
  };
}

File: back/persistence/wire/src/main/kotlin/EntityPayload.kt
package persistence.changes

import id.toId
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import persistence.model.AttendanceEmailRequest
import persistence.model.BasketExchange
import persistence.model.Contract
import persistence.model.DeliveryContract
import persistence.model.DeliveryTemplate
import persistence.model.DeviceToken
import persistence.model.EntityType
import persistence.model.ErrorReport
import persistence.model.Member
import persistence.model.MemberInvitation
import persistence.model.MemberJoinRequest
import persistence.model.Notification
import persistence.model.Organization
import persistence.model.OrganizationRequest
import persistence.model.Owner
import persistence.model.OwnerInvitation
import persistence.model.Producer
import persistence.model.ProducerAccount
import persistence.model.ProducerRequest
import persistence.model.ProductType

/**
 * Polymorphic carrier for the typed body of an entity in the sync protocol.
 *
 * The discriminator value matches the corresponding [EntityType] constant.
 */
@Serializable
sealed interface EntityPayload {
    val entityType: EntityType

    /**
     * Rewrites `tmp_*` FK references using the provided [tmpIdMap].
     * The default implementation returns `this` unchanged.
     * Override on payloads that embed FK references to other entities that may
     * carry `tmp_*` ids allocated earlier in the same sync batch.
     */
    fun rewriteTmpIds(tmpIdMap: Map<String, String>): EntityPayload = this

    /**
     * Extracts the entity's own `tmp_*` id when this payload represents a new
     * entity creation (i.e. the entity id starts with [ClientMutation.TMP_ID_PREFIX]).
     * Returns `null` for payloads that never carry `tmp_*` own ids, or when the
     * id is a real server-allocated id.
     */
    fun extractTmpId(): String? = null
}

@Serializable
@SerialName("ProductType")
data class ProductTypePayload(
    val productType: ProductType,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.ProductType

    override fun extractTmpId(): String? = productType.productTypeId.id.takeIf { it.startsWith(ClientMutation.TMP_ID_PREFIX) }
}

@Serializable
@SerialName("Organization")
data class OrganizationPayload(
    val organization: Organization,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.Organization

    override fun rewriteTmpIds(tmpIdMap: Map<String, String>): EntityPayload {
        if (tmpIdMap.isEmpty()) return this
        val org = organization
        val rewrittenProducers =
            org.producers.map { producer ->
                tmpIdMap[producer.producerAccountId.id]
                    ?.let { producer.copy(producerAccountId = it.toId()) }
                    ?: producer
            }
        val rewrittenProducts =
            org.products.map { product ->
                val newProducerAccountId =
                    tmpIdMap[product.producerAccountId.id]?.toId()
                        ?: product.producerAccountId
                val newProductTypeId =
                    tmpIdMap[product.productTypeId.id]?.toId()
                        ?: product.productTypeId
                if (newProducerAccountId != product.producerAccountId || newProductTypeId != product.productTypeId) {
                    product.copy(producerAccountId = newProducerAccountId, productTypeId = newProductTypeId)
                } else {
                    product
                }
            }
        val rewrittenDefaultTemplateId =
            org.defaultDeliveryTemplateId?.let { templateId ->
                tmpIdMap[templateId.id]?.toId() ?: templateId
            }
        val rewrittenDeliveries =
            org.deliveries.map { delivery ->
                val newTemplateId =
                    delivery.deliveryTemplateId?.let { templateId ->
                        tmpIdMap[templateId.id]?.toId() ?: templateId
                    }
                val rewrittenContracts =
                    delivery.contracts.map { contract ->
                        tmpIdMap[contract.contractId.id]
                            ?.let { contract.copy(contractId = it.toId()) }
                            ?: contract
                    }
                val rewrittenBasketDescriptions =
                    delivery.basketDescriptions.map { desc ->
                        tmpIdMap[desc.productTypeId.id]
                            ?.let { desc.copy(productTypeId = it.toId()) }
                            ?: desc
                    }
                if (
                    newTemplateId != delivery.deliveryTemplateId ||
                    rewrittenContracts != delivery.contracts ||
                    rewrittenBasketDescriptions != delivery.basketDescriptions
                ) {
                    delivery.copy(
                        deliveryTemplateId = newTemplateId,
                        contracts = rewrittenContracts,
                        basketDescriptions = rewrittenBasketDescriptions,
                    )
                } else {
                    delivery
                }
            }
        if (
            rewrittenProducers == org.producers &&
            rewrittenProducts == org.products &&
            rewrittenDefaultTemplateId == org.defaultDeliveryTemplateId &&
            rewrittenDeliveries == org.deliveries
        ) {
            return this
        }
        return copy(
            organization =
                org.copy(
                    producers = rewrittenProducers,
                    products = rewrittenProducts,
                    defaultDeliveryTemplateId = rewrittenDefaultTemplateId,
                    deliveries = rewrittenDeliveries,
                ),
        )
    }
}

@Serializable
@SerialName("ProducerAccount")
data class ProducerAccountPayload(
    val producerAccount: ProducerAccount,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.ProducerAccount

    override fun extractTmpId(): String? = producerAccount.producerAccountId.id.takeIf { it.startsWith(ClientMutation.TMP_ID_PREFIX) }

    override fun rewriteTmpIds(tmpIdMap: Map<String, String>): EntityPayload {
        if (tmpIdMap.isEmpty()) return this
        val rewrittenProducts =
            producerAccount.products.map { product ->
                tmpIdMap[product.productTypeId.id]
                    ?.let { product.copy(productTypeId = it.toId()) }
                    ?: product
            }
        return if (rewrittenProducts == producerAccount.products) {
            this
        } else {
            copy(producerAccount = producerAccount.copy(products = rewrittenProducts))
        }
    }
}

@Serializable
@SerialName("Member")
data class MemberPayload(
    val member: Member,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.Member
}

@Serializable
@SerialName("MemberJoinRequest")
data class MemberJoinRequestPayload(
    val memberJoinRequest: MemberJoinRequest,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.MemberJoinRequest
}

@Serializable
@SerialName("MemberInvitation")
data class MemberInvitationPayload(
    val memberInvitation: MemberInvitation,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.MemberInvitation

    override fun extractTmpId(): String? = memberInvitation.invitationId.takeIf { it.startsWith(ClientMutation.TMP_ID_PREFIX) }
}

@Serializable
@SerialName("Contract")
data class ContractPayload(
    val contract: Contract,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.Contract

    override fun extractTmpId(): String? = contract.contractId.id.takeIf { it.startsWith(ClientMutation.TMP_ID_PREFIX) }
}

@Serializable
@SerialName("DeliveryTemplate")
data class DeliveryTemplatePayload(
    val deliveryTemplate: DeliveryTemplate,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.DeliveryTemplate

    override fun extractTmpId(): String? = deliveryTemplate.deliveryTemplateId.id.takeIf { it.startsWith(ClientMutation.TMP_ID_PREFIX) }
}

@Serializable
@SerialName("OrganizationRequest")
data class OrganizationRequestPayload(
    val organizationRequest: OrganizationRequest,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.OrganizationRequest
}

@Serializable
@SerialName("ProducerRequest")
data class ProducerRequestPayload(
    val producerRequest: ProducerRequest,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.ProducerRequest
}

@Serializable
@SerialName("Owner")
data class OwnerPayload(
    val owner: Owner,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.Owner
}

@Serializable
@SerialName("OwnerInvitation")
data class OwnerInvitationPayload(
    val ownerInvitation: OwnerInvitation,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.OwnerInvitation

    override fun extractTmpId(): String? = ownerInvitation.invitationId.id.takeIf { it.startsWith(ClientMutation.TMP_ID_PREFIX) }
}

@Serializable
@SerialName("BasketExchange")
data class BasketExchangePayload(
    val basketExchange: BasketExchange,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.BasketExchange

    override fun extractTmpId(): String? = basketExchange.basketExchangeId.id.takeIf { it.startsWith(ClientMutation.TMP_ID_PREFIX) }

    override fun rewriteTmpIds(tmpIdMap: Map<String, String>): EntityPayload {
        if (tmpIdMap.isEmpty()) return this
        val exchange = basketExchange
        val newContractId =
            tmpIdMap[exchange.contractId.id]?.toId<DeliveryContract>()
                ?: exchange.contractId
        return if (newContractId != exchange.contractId) {
            copy(basketExchange = exchange.copy(contractId = newContractId))
        } else {
            this
        }
    }
}

@Serializable
@SerialName("Notification")
data class NotificationPayload(
    val notification: Notification,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.Notification
}

@Serializable
@SerialName("DeviceToken")
data class DeviceTokenPayload(
    val deviceToken: DeviceToken,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.DeviceToken

    override fun extractTmpId(): String? = deviceToken.deviceTokenId.id.takeIf { it.startsWith(ClientMutation.TMP_ID_PREFIX) }
}

@Serializable
@SerialName("AttendanceEmailRequest")
data class AttendanceEmailRequestPayload(
    val attendanceEmailRequest: AttendanceEmailRequest,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.AttendanceEmailRequest

    override fun extractTmpId(): String? =
        attendanceEmailRequest.attendanceEmailRequestId.id.takeIf {
            it.startsWith(ClientMutation.TMP_ID_PREFIX)
        }
}

@Serializable
@SerialName("Producer")
data class ProducerPayload(
    val producer: Producer,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.Producer
    // producerId is always the auth sub — never a tmp_* client-generated id
}

@Serializable
@SerialName("ErrorReport")
data class ErrorReportPayload(
    val errorReport: ErrorReport,
) : EntityPayload {
    override val entityType: EntityType get() = EntityType.ErrorReport

    override fun extractTmpId(): String? = errorReport.errorReportId.id.takeIf { it.startsWith(ClientMutation.TMP_ID_PREFIX) }
}
